### DISCUSSION TITLE
Promote Versions trough folders

### DISCUSSION THREAD

**Discussion by TheKangaroo** [renovatebot/renovate#27871] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

We have lots of platform repos with version pind in different folders for different stages. An example folder structure could look like this:
```
.
├── environments
│   ├── dev
│   │   ├── stage.tfvars
│   │   └── terragrunt.hcl
│   ├── prod
│   │   ├── stage.tfvars
│   │   └── terragrunt.hcl
│   └── terragrunt.hcl
├── renovate.json5
```

We pin version in both dev and prod folders and use the renovate regex manager to bump dependencies pinnend in both folders.
The config file looks like this in both dev and prod:

```yaml
[...]
# renovate: datasource=helm depName=ingress-nginx registryUrl=https://kubernetes.github.io/ingress-nginx
ingress_nginx_version: "1.2.3"
[...]
```
To be able to bump versions independently in dev and later in prod, we use the `additionalBranchPrefix' to split the bump into two separate PRs. 
This works fine so far, but if we bump a version in dev and a new version is available before we merge it into prod, we will get the PR for prod updated as well, but we didn't test that version in dev before.
For example
1) dependency1 1.0.0 is released
2) we merge 1.0.0 in dev, wait a few days for prod
3) dependency1 1.0.1 is released
4) renovate opens dev PR for 1.0.1 and updates prod PR from 1.0.0 to 1.01
5) We cannot merge the version tested in dev into prod without a manual commit.

What we really want, however, is for dev to point to the upstream source (as in the previous example), but for prod to point to the version already deployed in dev. 
I have no idea how to configure renovate to achieve this, or even if it is possible, but I thought this was something many users do.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by mrclrchtr** [comment]
You maybe can use `matchFileNames` to achieve this. I used it here: https://github.com/renovatebot/renovate/discussions/28272

**Reply by TheKangaroo** [reply]
Clever approach, but we already achieve the same result with `additionalBranchPrefix`, which splits the PRs into two because of the different branch names.
What we are additionally trying to achieve is that prod depends on the version we are running in dev and not on the upsream itself.
Thanks for the hint though :)

**Reply by mrclrchtr** [reply]
Maybe this is something that might interest you: https://github.com/akuity/kargo but I'm not entirely sure if I understood the requirement correctly ;)

**Comment by TheKangaroo** [comment, marked as answer]
Okay, I figured this one out. Shoutout to this [blog post](https://secustor.dev/blog/renovate_custom_datasources/) that pointed me in the right direction. It still took me a few hours to figure out how everything works, so here is the important part of my config:
```

  customManagers: [
    {
      customType: "regex",
      fileMatch: [".y[a]?ml$"],
      matchStrings: [
        '# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*?) registryUrl=(?<registryUrl>.*?)\\s.*_version:\\s"(?<currentValue>.*?)"\\s',
      ],
    },
  ],
  customDatasources: {
    local: {
      format: "yaml",
      transformTemplates: [
        '{ "releases": [{"version": `{{{replace "-" "_" packageName}}}_version`}]}',
      ],
    },
  },
```
(In the meantime, I migrated to .yaml config files, which made things easier for `customDatasource` lookup.)

Now I can specify  versions in `prod` dependant on `dev`
```
# renovate: datasource=custom.local depName=ingress-nginx registryUrl=file://environments/dev/config.yaml
ingress_nginx_version  = 1.2.3
```

**Reply by HonkingGoose** [reply]
@secustor you're getting a compliment in the comment above mine. :wink:
